### PR TITLE
WritePrepared Txn: Fix bug with duplicate keys during recovery

### PR DESCRIPTION
**Pull request by maysamyabandeh**
Fix the following bugs:
- During recovery a duplicate key was inserted twice into the write batch of the recovery transaction, 
once when the memtable returns false (because it was duplicates) and once for the 2nd attempt. This would result into different SubBatch count measured when the recovered transactions is committing.
- If a cf is flushed during recovery the memtable is not available to assist in detecting the duplicate key. This could result into not advancing the sequence number when iterating over duplicate keys of a flushed cf and hence inserting the next key with the wrong sequence number.
- SubBacthCounter would reset the comparator to default comparator after the first duplicate key. The 2nd duplicate key hence would have gone through a wrong comparator and not being detected.

### CODE REVIEW

File: db/write_batch.cc [hunk 1]
```diff
@@ -1135,17 +1190,23 @@ class MemTableInserter : public WriteBatch::Handler {
...
-        return Status::OK();
-      }
+      return Status::OK();
       // else insert the values to the memtable right away
     }
 
     Status seek_status;
-    if (!SeekToColumnFamily(column_family_id, &seek_status)) {
-      MaybeAdvanceSeq();
+    if (UNLIKELY(!SeekToColumnFamily(column_family_id, &seek_status))) {
```
Comment (yiwu-arbug) [inline db/write_batch.cc]: move line 1201-1209 and 1279-1283 close together and add comment to explain both of them are to handle flushed CFs on recovery?

File: db/db_impl.h [hunk 2]
```diff
@@ -550,9 +550,18 @@ class DBImpl : public DB {
     WriteBatch* batch_;
     // The seq number of the first key in the batch
     SequenceNumber seq_;
+    // Number of sub-batched. A new sub-batch is created if we txn attempts to
+    // inserts a duplicate key,seq to memtable. This is currently used in
+    // WritePrparedTxn
+    size_t batch_cnt_;
```
Comment (yiwu-arbug) [inline db/db_impl.h]: where is this being used?

File: db/db_impl.h [hunk 3]
```diff
@@ -550,9 +550,18 @@ class DBImpl : public DB {
...
     explicit RecoveredTransaction(const uint64_t log, const std::string& name,
-                                  WriteBatch* batch, SequenceNumber seq)
-        : log_number_(log), name_(name), batch_(batch), seq_(seq) {}
+                                  WriteBatch* batch, SequenceNumber seq,
+                                  size_t batch_cnt)
+        : log_number_(log),
+          name_(name),
+          batch_(batch),
+          seq_(seq),
+          batch_cnt_(batch_cnt) {}
```
Comment (yiwu-arbug) [inline db/db_impl.h]: Is this used only at the assert at pessimistic_transaction_db.cc:145-149?

File: db/db_impl.h (hunk 2 above)
Comment (maysamyabandeh) [reply db/db_impl.h]: utilities/transactions/pessimistic_transaction_db.cc:147 in an assert statement.

File: db/db_impl.h (hunk 3 above)
Comment (maysamyabandeh) [reply db/db_impl.h]: true

File: db/write_batch.cc [hunk 4]
```diff
@@ -978,6 +978,59 @@ Status WriteBatch::PopSavePoint() {
...
+      return true;
+    }
+    return false;
+  }
+
+ private:
+  SequenceNumber batch_seq_ = 0;
+  DBImpl* db_;
+  // A comparator to be used in std::set
+  struct SetComparator {
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply db/write_batch.cc]: i can add a TODO for that. Any class in util you have in mind?

File: db/write_batch.cc (hunk 1 above)
Comment (maysamyabandeh) [reply db/write_batch.cc]: i can add comments but not sure how to put them close together.

**Pull request closed without merging**